### PR TITLE
Adding flexible timesteps

### PR DESCRIPTION
**Pull request by jmorrisnrel**
Adding the option for Engage to support user defined timesteps on runs

Moving resampling from timeseries creation to run creation to support new timestep flexibility

### CODE REVIEW

File: calliope_app/api/views/outputs.py [hunk 1]
```diff
@@ -55,6 +55,14 @@ def build(request):
     end_date = request.GET.get("end_date", None)
     cluster = (request.GET.get("cluster", 'true') == 'true')
     manual = (request.GET.get("manual", 'false') == 'true')
+    timestep = request.GET.get("timestep", '1H')
+    try:
+        pd.tseries.frequencies.to_offset(timestep)
+    except ValueError:
+        payload = {
+            "status": "Failed",
+            "message": "'"+timestep+"' is not a valid timestep.",
```
Comment (jgu2) [inline calliope_app/api/views/outputs.py:64]: Return information about valid `timestep` options?

File: calliope_app/api/tasks.py [hunk 2]
```diff
@@ -319,8 +319,18 @@ def get_timeseries_data(filename, start_date, end_date):
...
-    idx = pd.period_range(start_date, end_date, freq='H').to_timestamp()
+    # Determine offset difference and resample to run's timesteps
+    diffs = (subset.index[1:]-subset.index[:-1])
+    min_delta = diffs.min()
+    subset_freq = start_date+min_delta
+    target_freq = start_date+pd.tseries.frequencies.to_offset(timesteps)
+    if subset_freq<target_freq:
+        subset = subset.resample(rule=timesteps).mean()
+    elif subset_freq>target_freq:
+        subset = subset.resample(rule=timesteps).interpolate()
```
[the remark this answers is not in the document]
Comment (jmorrisnrel) [reply calliope_app/api/tasks.py:330]: Yes but there's no need to resample if the frequencies are equal so no need for an `else` I think

File: calliope_app/api/tasks.py [hunk 3]
```diff
@@ -319,8 +319,18 @@ def get_timeseries_data(filename, start_date, end_date):
     subset = subset[~feb_29_mask]
     subset.index = subset.index.map(lambda t: t.replace(year=year))
 
-    # Fill Missing Hours w/ 0
-    idx = pd.period_range(start_date, end_date, freq='H').to_timestamp()
+    # Determine offset difference and resample to run's timesteps
+    diffs = (subset.index[1:]-subset.index[:-1])
```
[the remark this answers is not in the document]
Comment (jmorrisnrel) [reply calliope_app/api/tasks.py:323]: Around the `subset` line? Rob had those in there for whatever reason. I'll remove them

**Review by jgu2 — APPROVED** [review]
No critical to change, approved.

**Pull request merged**